tests/unit-tests/test-class-sensei-course-enrolment.php

```diff
@@ -0,0 +1,385 @@
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Always_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertTrue( $course_enrolment->is_enroled( $student_id ), 'One positive provider should always enrol student in course' );
+	}
+
+	/**
+	 * Check for negative result if no providers that handle this course's registration are handled.
+	 */
+	public function testEnrolmentCheckNoHandlingProviders() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Handles::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id ), 'Student should not be enroled if there are no providers handling enrolment for a course' );
+	}
+
+	/**
+	 * Tests to make sure non-handling providers aren't used. Never_Handles has a true response, but it should never be asked.
+	 */
+	public function testEnrolmentCheckNonHandlingProvidersIgnored() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Handles::class );
+		$this->addEnrolmentProvider( Sensei_Test_Enrolment_Provider_Never_Provides::class );
+
+		$course_enrolment = Sensei_Course_Enrolment::get_course_instance( $course_id );
+
+		$this->assertFalse( $course_enrolment->is_enroled( $student_id ), 'Non-handling providers should never be used to determine enrolment' );
+	}
+
+	/**
+	 * Tests to make sure non-handling providers aren't used. Never_Handles has a true response, but it should never be asked.
+	 */
+	public function testEnrolmentCheckVersionCachingWorks() {
+		$course_id  = $this->getSimpleCourse();
+		$student_id = $this->createStandardStudent();
+
+		// Initial request.
+		$this->resetAndSetUpVersiondProvider( false ); // Version : 1; Odd numbers do not provide enrolment.
+		$course_enrolment_a = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$is_enroled_a       = $course_enrolment_a->is_enroled( $student_id );
+		$is_enroled_a_2     = $course_enrolment_a->is_enroled( $student_id );
+		$this->assertFalse( $is_enroled_a, 'This version of the provider should not provide enrolment' );
+		$this->assertEquals( $is_enroled_a_2, $is_enroled_a, 'Duplicate calls should use caching and return same result' );
+
+		// Bump version. Second request.
+		$this->resetAndSetUpVersiondProvider( true ); // Version : 2; Even numbers provide enrolment.
+		$course_enrolment_b = Sensei_Course_Enrolment::get_course_instance( $course_id );
+		$is_enroled_b       = $course_enrolment_b->is_enroled( $student_id );
+		$this->assertTrue( $is_enroled_b, 'This version of the provider should provide enrolment. Cache was not refreshed on version change' );
```

Comment:
Like the previous comment, what do you think about to use a description like this (or something similar)?
`Cache should refresh on version change. And this version of the provider should provide enrolment.`

Reply:
0c80b60fbf0e34a8e5eb5c2c4124ba00f7021d0e